### PR TITLE
Handle status properly

### PR DESCRIPTION
**Pull request by barchw**
**Description**

Changes proposed in this pull request:

- Set error status on the resource on which the error occurred

**Related issue(s)**

- https://github.com/kyma-project/api-gateway/issues/84

### CODE REVIEW

File: internal/processing/reconciliation_test.go [hunk 1]
```diff
@@ -140,12 +168,54 @@ var _ = Describe("Reconcile", func() {
...
+		scheme := runtime.NewScheme()
+		err := networkingv1beta1.AddToScheme(scheme)
+		Expect(err).NotTo(HaveOccurred())
+		client := fake.NewClientBuilder().WithScheme(scheme).Build()
+
+		// when
+ 		status := processing.Reconcile(context.TODO(), client, testLogger(), cmd, &gatewayv1beta1.APIRule{})
+
+		// then
+		Expect(status.ApiRuleStatus.Code).To(Equal(gatewayv1beta1.StatusOK))
```
Comment (triffer) [inline internal/processing/reconciliation_test.go:206]: I wonder if the ApiRule status should be ok if the status of VS is not. Shouldn't we reflect the sub resource status in the ApiRule as this is the user-facing configuration?

**Review by triffer — COMMENTED** [review]
I think we should evaluate if we need to add assertions for the new status handling to existing tests in `api_controller_integration_test.go`. Besides this I think we should add dedicated status integration test for each handler, because the current tests don't cover this as they have the base status mocked.

File: api/v1beta1/apirule_types.go [hunk 2]
```diff
@@ -45,20 +45,23 @@ type APIRuleSpec struct {
...
-	LastProcessedTime           *metav1.Time           `json:"lastProcessedTime,omitempty"`
-	ObservedGeneration          int64                  `json:"observedGeneration,omitempty"`
-	APIRuleStatus               *APIRuleResourceStatus `json:"APIRuleStatus,omitempty"`
-	VirtualServiceStatus        *APIRuleResourceStatus `json:"virtualServiceStatus,omitempty"`
-	AccessRuleStatus            *APIRuleResourceStatus `json:"accessRuleStatus,omitempty"`
+	LastProcessedTime    *metav1.Time           `json:"lastProcessedTime,omitempty"`
+	ObservedGeneration   int64                  `json:"observedGeneration,omitempty"`
+	APIRuleStatus        *APIRuleResourceStatus `json:"APIRuleStatus,omitempty"`
+	VirtualServiceStatus *APIRuleResourceStatus `json:"virtualServiceStatus,omitempty"`
+	// +optional
```
[the remark this answers is not in the document]
Comment (barchw) [reply api/v1beta1/apirule_types.go:59]: The field is already optional as it has "omitempty", just making it more verbose here

File: internal/processing/helpers.go [hunk 3]
```diff
@@ -98,3 +100,23 @@ func SliceToString(ss []string) (s string) {
 	}
 	return
 }
+
+func GenerateValidationDescription(failures []validation.Failure) string {
```
[the remark this answers is not in the document]
Comment (barchw) [reply internal/processing/helpers.go:104]: Hmm, I wonder why lint didn't catch it

File: internal/processing/reconciliation_test.go (hunk 1 above)
Comment (barchw) [reply internal/processing/reconciliation_test.go:206]: I will use Error on APIRule also

File: internal/processing/status.go [hunk 4]
```diff
@@ -15,52 +15,71 @@ type ReconciliationStatus struct {
...
+	case OnVirtualService:
+		return "VirtualService"
+	case OnAccessRule:
+		return "Rule"
+	case OnRequestAuthentication:
+		return "RequestAuthentication"
+	case OnAuthorizationPolicy:
+		return "AuthorizationPolicy"
+	default:
+		return "APIRule"
```
Comment (barchw) [inline internal/processing/status.go:39]:
```suggestion
                // If no Kind is resolved from the resource (e.g. subresource CRD is missing)
		return "APIRule"
```

File: internal/processing/reconciliation_test.go [hunk 5]
```diff
@@ -104,8 +126,11 @@ var _ = Describe("Reconcile", func() {
 		// then
 		Expect(status.ApiRuleStatus.Code).To(Equal(gatewayv1beta1.StatusError))
 		Expect(status.ApiRuleStatus.Description).ToNot(BeEmpty())
-		Expect(status.AccessRuleStatus.Code).To(Equal(gatewayv1beta1.StatusError))
-		Expect(status.VirtualServiceStatus.Code).To(Equal(gatewayv1beta1.StatusError))
+		Expect(status.AccessRuleStatus.Code).To(Equal(gatewayv1beta1.StatusOK))
+		Expect(status.VirtualServiceStatus.Code).To(Equal(gatewayv1beta1.StatusOK))
```
[the remark this answers is not in the document]
Comment (barchw) [reply internal/processing/reconciliation_test.go:130]: I will add info about VS kind missing

**Review by triffer — APPROVED** [review]

**Pull request merged**